llvm/lib/SYCLLowerIR/ESIMD/LowerESIMDKernelAttrs.cpp

```diff
@@ -30,7 +63,28 @@ SYCLFixupESIMDKernelWrapperMDPass::run(Module &M, ModuleAnalysisManager &MAM) {
       sycl::utils::traverseCallgraphUp(
           &F,
           [&](Function *GraphNode) {
-            if (!llvm::esimd::isESIMD(*GraphNode)) {
+            if (!llvm::esimd::isESIMD(*GraphNode) &&
+                llvm::esimd::isKernel(*GraphNode)) {
+
+              // Demangle the caller name to check if the function is called
+              // from RoundedRangeKernel.
+              StringRef MangledName = GraphNode->getName();
+              llvm::itanium_demangle::ManglingParser<SimpleAllocator> Parser(
+                  MangledName.begin(), MangledName.end());
+              itanium_demangle::Node *AST = Parser.parse();
+              if (!AST ||
+                  AST->getKind() != itanium_demangle::Node::KSpecialName) {
+                return;
+              }
+
+              itanium_demangle::OutputBuffer NameBuf;
+              AST->print(NameBuf);
+              StringRef Name(NameBuf.getBuffer(), NameBuf.getCurrentPosition());
+
+              if (!Name.contains("sycl::_V1::detail::RoundedRangeKernel<")) {
+                return;
+              }
+
```

Comment:
looks like the `GraphNode->setMetadata...` call below is reached only when GraphNode is a RoundedRangeKernel, which means only RoundedRangeKernel gets the marker, all others are left out. Previous logic was that all functions were marked. This needs additional investigation. I'll also give this a thought.

Reply:
What this patch does is it excludes all the functions in the caller chain from given ESIMD function F up to RoundedRangeKernel from marking them as ESIMD, instead, it only marks the kernel. Previously all the function in the call chain were marked.
Now, this patch is supposed to fix internal LLVM error on some function met in the IR:
Assertion `isa<To>(Val) && "cast<Ty>() argument of incompatible type!"' failed.

So
1) All the functions in a call chain ending at ESIMD function must be ESIMD functions, so not marking them is incorrect.
2) I think the patch just masks the problem by skipping marking the problematic function (which in turn can spawn other problems).

I think we need to come up with a patch that fixes the root cause of the above assert.